docs/proposals/edi.md

```diff
@@ -0,0 +1,258 @@
+to parse EDI files and write additional logic on those data.
+
+## Module Overview
+
+EDI module provides functionality to read EDI files and map those to Ballerina records or 'json' type. Mappings for EDI
+files have to be provided in json format. Once a mapping is provided, EDI module can generate Ballerina records to hold
+data in any EDI file represented by that mapping. Then the module can read EDI files (in text format) in to generated
+Ballerina records or as json values, which can be accessed from Ballerina code.
+
+## EDI file structure
+
+Here is a simple sample for EDI file
+
+```
+HDR*ORDER_1201*ABC_Store*2008-01-01~
+ITM*A-250*12
+ITM*A-45*100
+ITM*D-10*58
+ITM*K-80*250
+ITM*T-46*28
+```
+
+An EDI file is a binary computer file that contains data arranged in units called data elements. Data elements are
+separated by element terminators(*), and a group of data elements makes up a data segment. Data segments are separated
+by
+segment terminators(~).
+
+## EDI mapping file
+
+In order to parse an EDI document to ballerina readable code, an EDI mapping file is required. The mapping file can be
+manually implemented by a human for a given EDI document or can be converted from an existing definition like `Smooks`
+or `ESL`.
+
+Here is a sample EDI mapping file(edi-mapping1.json) for the above EDI document.
+
+```json
+{
+  "name": "SimpleOrder",
+  "delimiters": {
+    "segment": "~",
+    "field": "*"
+  },
+  "segments": {
+    "HDR": {
+      "tag": "header",
+      "fields": [
+        {
+          "tag": "orderId"
+        },
+        {
+          "tag": "organization"
+        },
+        {
+          "tag": "date"
+        }
+      ]
+    },
+    "ITM": {
+      "tag": "items",
+      "maxOccurances": -1,
+      "fields": [
+        {
+          "tag": "item"
+        },
+        {
+          "tag": "quantity",
+          "dataType": "int"
+        }
+      ]
+    }
+  }
+}
+```
+
+Above mapping can be used to parse EDI documents with one HDR segment (mapped to "header") and any number of ITM
+segments (mapped to "items"). HDR segment contains three fields, which are mapped to "orderId", "organization" and "
+date". Each ITM segment contains two fields mapped to "item" and "quantity". Below is a sample EDI document that can be
+parsed using the above mapping (let's assume that the below EDI is saved in edi-sample1.edi file).
+
+## Converting Smooks mapping files to Ballerina mappings
+
+Smooks library is commonly used for parsing EDI files. Therefore, many organizations have already created Smooks
+mappings for their EDIs. Ballerina EDI module can convert such Smooks mapping to Ballerina compatible mappings so that
+organizations can start using Ballerina for EDI processing without redoing any mappings.
+
+The following command converts Smooks EDI mapping to Ballerina EDI mapping:
+
+```shell
+bal smooksToBal <Smooks mapping xml file path> <Ballerina mapping json file path>
```

Comment:
IMO we better keep all the EDI sub-commands under the `bal edi` command (similar to what we have done for `bal dist` command).
Then the commands would look like below. 
 - `$bal edi smooks <args>`
 -  `$bal edi esl <args>`
 -  `$bal edi gen <args>`
 -  `$bal edi lib <args>`

Reply:
+1 for `$bal edi gen <args>` and `$bal edi lib <args>`.
But `$bal edi smooks <args>` and  `$bal edi esl <args>` do not imply what exactly those commands do.

Reply:
Yes, we may have to find better names for bal edi smook and esl. May be "bal edi smooks-convert ...", etc.
Anyway we have wait until module specific commands are supported in bal command.